File: 128. Longest Consecutive Sequence/solution.py
from typing import List


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        max_len = 0
        num_set = set(nums)
        for num in num_set:
            if num-1 in num_set:
                continue
            current_longest = 1
            while num+current_longest in num_set:
                current_longest += 1
            max_len = max(max_len, current_longest)
        return max_len

File: 128. Longest Consecutive Sequence/solution.go
package main

// Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.

// You must write an algorithm that runs in O(n) time.

func longestConsecutive(nums []int) int {
    // create a set to store non duplicate int
	numSet := make(map[int]struct{})
	for _, num := range nums {
		numSet[num] = struct{}{}
	}
	maxLen := 0
	for k, _ := range numSet {
		// we will look up, so if k-1 in set, it meas it has already been visited, we skip to reduce time complexity
		if _, exists := numSet[k-1]; exists {
			continue
		}
		currMaxLen := 1
		for {
			if _, exists := numSet[k+currMaxLen]; exists {
				currMaxLen++
			} else {
				break
			}
		}
		maxLen = max(currMaxLen, maxLen)
	}
	return maxLen
}